### PR TITLE
Added support for Codeception test framework

### LINKED ISSUE #45: Integrate Codeception testing framework
I'm using [codeception](http://codeception.com/) as my main testing framework and may be interested in providing the integration. Has anybody else already started working on codeception integration or have there been any blockers I don't know about yet?

### PR DESCRIPTION
**Pull request by tobiasstadler**
Fixes #45

### CODE REVIEW

**Comment by DavertMik** [comment]
Any updates?

**Comment by theofidry** [comment]
Will try to review that in the course of next week, I'm sure the rest of the team will too. Thanks for the hard work and sorry to be a bit slow on this one :)

File: src/TestFramework/Codeception/CommandLine/ArgumentsAndOptionsBuilder.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+    public function build(string $configPath, string $extraOptions, Mutant $mutant = null): string
+    {
+        $options = [
+            'run',
+            '--no-colors'
+        ];
+
+        if ($mutant !== null) {
+            $options[] = '-o "paths: output: ' . $this->tempDir . '/' . $mutant->getMutation()->getHash() . '"';
```
Comment (borNfreee) [inline src/TestFramework/Codeception/CommandLine/ArgumentsAndOptionsBuilder.php]: could you please explain what is it used for?

File: src/TestFramework/Codeception/Config/Builder/InitialConfigBuilder.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    private $originalConfigPath;
+
+    public function __construct(string $originalConfigPath)
+    {
+        $this->originalConfigPath = $originalConfigPath;
+    }
+
+    public function build(): string
+    {
+        return $this->originalConfigPath;
```
Comment (borNfreee) [inline src/TestFramework/Codeception/Config/Builder/InitialConfigBuilder.php]: Maybe it's better to create all needed for Infection parameters in a new, temp config like it is done for PHPUnit and PHPSpec rather that adding/overriding config in `ArgumentsAndOptionsBuilder`?

This will probably help to avoid passing `Mutant $mutant` parameter to `ArgumentsAndOptionsBuilder`

In other words, I suggest to move this code:

```php
$options[] = '-o "paths: output: ' . $this->tempDir . '"';
$options[] = '-o "coverage: enabled: true"';
$options[] = '--coverage-phpunit ' . CodeCoverageData::CODECEPTION_COVERAGE_DIR;
```

from `ArgumentsAndOptionsBuilder` and create temp config. It will be consistent with the other Test Framework and will help to avoid changing the interface

File: src/TestFramework/Codeception/Config/Builder/MutationConfigBuilder.php [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        return $this->originalConfigPath;
+    }
+
+    private function createCustomAutoloadWithInterceptor(Mutant $mutant): string
+    {
+        $originalFilePath = $mutant->getMutation()->getOriginalFilePath();
+        $mutatedFilePath = $mutant->getMutatedFilePath();
+        $interceptorPath = dirname(__DIR__, 4) . '/StreamWrapper/IncludeInterceptor.php';
+
+        // TODO change to what it was (e.g. app/autoload - see simplehabits)
```
Comment (borNfreee) [inline src/TestFramework/Codeception/Config/Builder/MutationConfigBuilder.php]: please do not copy todo comments

File: src/TestFramework/Codeception/Config/Builder/MutationConfigBuilder.php [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    public function __construct(string $tempDirectory, string $projectDirectory, string $originalConfigPath)
+    {
+        $this->tempDirectory = $tempDirectory;
+        $this->projectDirectory = $projectDirectory;
+        $this->originalConfigPath = $originalConfigPath;
+    }
+
+    public function build(Mutant $mutant): string
+    {
+        $_SERVER['CUSTOM_AUTOLOAD_FILE_PATH'] = $customAutoloadFilePath = sprintf(
```
Comment (borNfreee) [inline src/TestFramework/Codeception/Config/Builder/MutationConfigBuilder.php]: could you please expand on why do you need it?

File: src/TestFramework/Codeception/CustomAutoloadFilePath.php [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ * Copyright © 2018 Tobias Stadler
+ *
+ * License: https://opensource.org/licenses/BSD-3-Clause New BSD License
+ */
+
+declare(strict_types=1);
+
+namespace Infection\TestFramework\Codeception;
+
+use Codeception\Events;
```
Comment (borNfreee) [inline src/TestFramework/Codeception/CustomAutoloadFilePath.php]: 1. it will not work since `codeception/codeception` is in `require-dev` section and all files from this section are absent in PHAR distribution
2. we need to find a way to not use Codeception dependency at all. Infection should not depend on any test framework.

could you please explain what is the goal of this file, we will try to find another way to achieve the same goal

File: tests/TestFramework/Codeception/Adapter/CodeceptionAdapterTest.php [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        return [
+            ['OK, but incomplete, skipped, or risky tests!', true],
+            ['OK (5 tests, 3 assertions)', true],
+            ['FAILURES!', false],
+            ['ERRORS!', false],
+        ];
+    }
+
+    protected function tearDown()
```
Comment (borNfreee) [inline tests/TestFramework/Codeception/Adapter/CodeceptionAdapterTest.php]: please extend `Mockery\Adapter\Phpunit\MockeryTestCase` file. No need to use this tearDown function anymore

File: phpstan.neon [hunk 7]
```diff
@@ -1,4 +1,5 @@
 parameters:
     ignoreErrors:
         - '#Access to an undefined property PhpParser\\Node(.*?)#'
-        - '#Call to an undefined method PhpParser\\Node(.*?)#'
\ No newline at end of file
+        - '#Call to an undefined method PhpParser\\Node(.*?)#'
+        - '#Call to an undefined method Mockery\\MockInterface(.*?)#'
```
[the remark this answers is not in the document]
Comment (tobiasstadler) [reply phpstan.neon]: phpstan complains about a call to an undefined method Mockery\MockInterface::getMutation() in tests/TestFramework/Codeception/CommandLine/ArgumentsAndOptionsBuilderTest.php

File: src/TestFramework/Codeception/Config/Builder/InitialConfigBuilder.php (hunk 2 above)
Comment (tobiasstadler) [reply src/TestFramework/Codeception/Config/Builder/InitialConfigBuilder.php]: unfortunately Codeception does not handle absolute paths in its config files, but maybe I can get it working using relative paths.

File: src/TestFramework/Factory.php [hunk 8]
```diff
@@ -104,10 +108,22 @@ public function create(string $adapterName): AbstractTestFrameworkAdapter
...
+                new CodeceptionArgumentsAndOptionsBuilder($this->tempDir),
+                $this->versionParser
+            );
+        }
+
         throw new \InvalidArgumentException(
             sprintf(
                 'Invalid name of test framework. Available names are: %s',
-                implode(', ', [TestFrameworkTypes::PHPUNIT, TestFrameworkTypes::PHPSPEC])
+                implode(', ', [TestFrameworkTypes::PHPUNIT, TestFrameworkTypes::PHPSPEC, TestFrameworkTypes::CODECEPTION])
```
Comment (sidz) [inline src/TestFramework/Factory.php]: could you please use `TestFrameworkTypes::TYPES` constant here?

File: src/Command/InfectionCommand.php [hunk 9]
```diff
@@ -69,7 +69,7 @@ protected function configure()
                 'test-framework',
                 null,
                 InputOption::VALUE_REQUIRED,
-                'Name of the Test framework to use (phpunit, phpspec)',
+                'Name of the Test framework to use (phpunit, phpspec, codeception)',
```
Comment (sidz) [inline src/Command/InfectionCommand.php]: could you please use `TestFrameworkTypes::TYPES` constant here as we do in the ConfigureCommand?

File: tests/Fixtures/e2e/Codeception_Basic/run_test.bash [hunk 10]
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env bash
+if [[ $PHPDBG = 1 ]]
```
Comment (sidz) [inline tests/Fixtures/e2e/Codeception_Basic/run_test.bash]: @borNfreee told that he had an issue with this variant on his mac. Could you please change this line to the `if [ "$PHPDBG" = "1" ]` ?

Comment (tobiasstadler) [reply tests/Fixtures/e2e/Codeception_Basic/run_test.bash]: I had an issue with if [[ $PHPDBG=1 ]] on my Mac.  if [[ $PHPDBG = 1 ]] works for me.

@borNfreee Does it also work for you?

Comment (borNfreee) [reply tests/Fixtures/e2e/Codeception_Basic/run_test.bash]: yes, it does not work, see https://github.com/infection/infection/pull/151#discussion_r166079227

File: src/TestFramework/Codeception/Config/Builder/InitialConfigBuilder.php (hunk 2 above)
Comment (DavertMik) [reply src/TestFramework/Codeception/Config/Builder/InitialConfigBuilder.php]: I'd suggest creating Codeception's Extension which will transform the current config the way you need it. 
Extension can be published as a standalone package

File: src/TestFramework/Codeception/Config/Builder/MutationConfigBuilder.php (hunk 4 above)
Comment (tobiasstadler) [reply src/TestFramework/Codeception/Config/Builder/MutationConfigBuilder.php]: This is needed in order to autoload the mutated file. As far as I know Codeception does not support a custom bootstrap file (as e.g. PHPUnit)

File: src/TestFramework/Codeception/CustomAutoloadFilePath.php (hunk 5 above)
Comment (tobiasstadler) [reply src/TestFramework/Codeception/CustomAutoloadFilePath.php]: I don't thin this is a problem, because the code path is only used when the test framework is codeception.

File: tests/e2e_tests [hunk 11]
```diff
@@ -19,7 +19,7 @@ do
         composer install > /dev/null 2>&1
     fi
 
-    if [ -f "run_tests.bash" ]
+    if [ -f "run_test.bash" ]
```
Comment (BackEndTea) [inline tests/e2e_tests]: If you rebase on master please undo this change.

The intended filename is `run_tests.bash`, the `then` clause not running the correct file has been fixed on master.

File: tests/Fixtures/e2e/Codeception_Basic/run_test.bash [hunk 12]
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env bash
```
Comment (BackEndTea) [inline tests/Fixtures/e2e/Codeception_Basic/run_test.bash]: Please rename the file to `run_tests.bash`

File: tests/Fixtures/e2e/Codeception_Basic/composer.json [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+      "psr-4": {
+        "Codeception_Basic\\": "src/"
+      }
+    },
+    "autoload-dev": {
+      "psr-4": {
+        "Codeception_Basic\\Test\\": "tests/unit/"
+      }
+    }
+  }
```
Comment (BackEndTea) [inline tests/Fixtures/e2e/Codeception_Basic/composer.json]: Missing a new line at the end of the file

File: tests/TestFramework/Codeception/Config/YamlConfigurationHelperTest.php [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+
+namespace Infection\Tests\TestFramework\Codeception\Config;
+
+use Infection\Filesystem\Filesystem;
+use Infection\TestFramework\Codeception\Config\YamlConfigurationHelper;
+use Infection\Utils\TmpDirectoryCreator;
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\Yaml\Yaml;
+
+function rp(string $path) : string
```
Comment (BackEndTea) [inline tests/TestFramework/Codeception/Config/YamlConfigurationHelperTest.php]: Could you move this function into the test class as a helper function

File: src/TestFramework/Codeception/Config/Builder/MutationConfigBuilder.php (hunk 4 above)
Comment (sidz) [reply src/TestFramework/Codeception/Config/Builder/MutationConfigBuilder.php]: please use `putenv` instead of touching global variables

File: src/TestFramework/AbstractTestFrameworkAdapter.php [hunk 15]
```diff
@@ -74,12 +74,12 @@ public function __construct(
      *
      * @return string
      */
-    public function getExecutableCommandLine(string $configPath, string $extraOptions, bool $includePhpArgs = true): string
+    public function getExecutableCommandLine(string $configPath, string $extraOptions, bool $includePhpArgs = true, Mutant $mutant = null): string
```
[the remark this answers is not in the document]
Comment (borNfreee) [reply src/TestFramework/AbstractTestFrameworkAdapter.php]: this still needs to be fixed

File: src/TestFramework/Codeception/Config/Builder/MutationConfigBuilder.php (hunk 4 above)
Comment (sidz) [reply src/TestFramework/Codeception/Config/Builder/MutationConfigBuilder.php]: Infection runs frameworks via symfony console which knows how to work with envs. We already use it :P.

p.s. IMHO Don't like to touch global variables

Comment (theofidry) [reply src/TestFramework/Codeception/Config/Builder/MutationConfigBuilder.php]: I'm not sure to follow, I don't see how using the Console component helps to make it thread-safe here.

Maybe for reference: https://github.com/symfony/recipes/pull/156

Also I don't like to touch global variables much, but that's exactly what put/getenv are doings and regardless of what we are using we should use a prefix here IMO

**Comment by BackEndTea** [comment]
Could you run `composer cs:fix`?

**Comment by BackEndTea** [comment]
I'll give this good look over after the weekend, and see if i can break anything.

File: src/TestFramework/Factory.php [hunk 16]
```diff
@@ -111,10 +115,23 @@ public function create(string $adapterName, bool $skipCoverage): AbstractTestFra
             );
         }
 
+        if ($adapterName === TestFrameworkTypes::CODECEPTION) {
+            $codeceptionConfigPath = $this->configLocator->locate(TestFrameworkTypes::CODECEPTION);
+            $codeceptionConfigContent = file_get_contents($codeceptionConfigPath);
+
+            return new CodeceptionAdapter(
+                new TestFrameworkExecutableFinder(CodeceptionAdapter::EXECUTABLE),
+                new CodeceptionInitialConfigBuilder($this->tempDir, $this->projectDir, $codeceptionConfigContent, $this->infectionConfig->getSourceDirs()),
```
Comment (Fenikkusu) [inline src/TestFramework/Factory.php]: I was just trying this out on a project. A recent MR has changed the variable `$this->tempDir` to `$this->tmpDir`. I had to fix this before I could use the PR.

File: src/TestFramework/Factory.php [hunk 17]
```diff
@@ -112,10 +116,23 @@ public function create(string $adapterName, bool $skipCoverage): AbstractTestFra
             );
         }
 
+        if ($adapterName === TestFrameworkTypes::CODECEPTION) {
+            $codeceptionConfigPath = $this->configLocator->locate(TestFrameworkTypes::CODECEPTION);
+            $codeceptionConfigContent = file_get_contents($codeceptionConfigPath);
+
+            return new CodeceptionAdapter(
+                new TestFrameworkExecutableFinder(CodeceptionAdapter::EXECUTABLE),
```
Comment (Fenikkusu) [inline src/TestFramework/Factory.php]: Not Found For Me. Changed It To `TestFrameworkFinder` and started working again.

**Comment by Fenikkusu** [comment]
So far so good. I'm having issues running it through a docker image ( #254 ), but I think that is related more to infection than the codeception work.

**Comment by sidz** [comment]
@Fenikkusu first of all this PR has issues which should be fixed. So @tobiasstadler could you please check Travis build https://travis-ci.org/infection/infection/builds/357027746?utm_source=github_status&utm_medium=notification

**Comment by Fenikkusu** [comment]
I've come across an issue I didn't come across originally. If you are using params:env:.env, The system chokes because it can't find the .env file. I'm working on resolving the issue and have a partial fix at https://github.com/twistersfury/infection:codeception. However, this resulted in additional errors I have not yet had the time to resolve.

**Comment by Fenikkusu** [comment]
I've figured out the additional errors. I was running into two issues. The first was I had an acceptance suite installed, but no actual tests installed. When coverage ran, codeception choked because it was trying to connect to the acceptance url, which didn't exist. This error was easy enough for me to resolve, just by disabling the suite. However, it brings up the point of should the ability be added as part of the configuration to be able to specify which suites are ran?

The next issue took me a bit to track down. This time, phpunit was choking on tokenization for coverage with phpunit. Basically I had files in the project that were empty, but weren't PHP files. I tracked down the command infection was running and tried to replicate the specifics. I ultimately realized that when the Codeception suite runs, it is setting the configuration of coverage files to '*' based off the infection config. IE:

```json
    "source": {
        "directories": [
            "src"
        ]
    }
```

Is translated into:
```yaml
coverage:
  enabled: true
  include:
    - src/*
```

Because it is '*', it is including all files, even non-php ones. I'm fairly certain if we just remove the '*' that will fix the issue. My question though, at this point is if there is a specific reason why * was used instead of just using the folder path?

**Comment by Fenikkusu** [comment]
> So for those suites inflection should be disabled, that's for sure.

I agree that they should be disabled, but I still think the ability needs to be passed onto the user via configuration. I say that because the suites are not limited to unit, functional, and acceptance. For example, I've a project that has a suite called 'legacy' and another called 'coverage'. Both of these suites run under the pretense of unit testing and should be ran as part of the testing. I think in the end, an option in the json configuration needs to be added for suites. When not specified, it defaults to 'unit'. When specified, it runs all the suites specified. This could easily be a string that just directly injects into the command line arguments.

**Comment by Fenikkusu** [comment]
Update: I tested the code using just the path and not the '*'. I still run into the same issue. The easiest solution is to just change it to be *.php. Done and done. The only problem I see with that is that the possibility of file names not ending in .php. IE: php5, php4, and phtml are all considered valid extensions, though rare (with phtml likely not being a class).

File: src/TestFramework/Codeception/CustomAutoloadFilePath.php (hunk 5 above)
Comment (sanmai) [reply src/TestFramework/Codeception/CustomAutoloadFilePath.php]: This could work if this file won't be a part of Infection. Instead you could tell Codeception to include a file from the outside Infection. E.g. there's an `IncludeInterceptor` which [is added to the tested code at one point](https://github.com/infection/infection/blob/5bfd1be14bd3762a573717e118f6115a3426d77f/src/TestFramework/PhpSpec/Config/Builder/MutationConfigBuilder.php#L70). I'm not entirely sure you could do what you want there, but there's an opportunity.

**Comment by Fenikkusu** [comment]
Noting Additional Issue: If codeception errors (IE: use db module without an accessible db), the error isn't getting properly echoed by infection.

**Comment by borNfreee** [comment]
> the error isn't getting properly echoed by infection

that sucks. Could you please try set error hanlder as it is done in https://github.com/infection/infection/pull/275 and see if it helps?

**Comment by Fenikkusu** [comment]
After a bit of digging, I was able to figure out how to get this to replicate consistently. If I'm not debugging, I never get any output from the error (again, use the db module with invalid db). If I put in a break point and force the codecept command to finish `before` the incept command has time to kick off the wait process, I get the proper output. AKA, break on line 303 of Symfony's Process.php file and step the codecept run to the next line. Let the codecept finish, then continue running infection. In this case, you will get the correct output.

**Comment by Fenikkusu** [comment]
> that sucks. Could you please try to set an error hanlder as it is done in #275 and see if it helps?

@borNfreee, I tried that as well but saw no difference. At this point it appears to be a race condition in the underlying symfony framework.

**Comment by BackEndTea** [comment]
@tobiasstadler  are you still working on this?

**Comment by tobiasstadler** [comment]
Sorry, I am not working on this anymore.

**Comment by BackEndTea** [comment]
> Sorry, I am not working on this anymore.

I'm gonna go ahead and close this PR then. Thank you for the work you put into it.
Even though this won't be merged it may be usefull if someone wants to pick it up at a later point.

**Comment by Fenikkusu** [comment]
I've already done number of updates to this, though I haven't refreshed it in a couple of months. I'll open a new MR once I have a chance to rebase and clean up. I've been using it in a few applications, so I think it's about 80% there.

**Comment by maks-rafalko** [comment]
Implemented in https://github.com/infection/infection/pull/800
